## **Business Understanding**

#### **Latar Belakang**
Penyakit jantung merupakan salah satu penyebab kematian tertinggi di dunia menurut World Health Organization. Salah satu indikator pending dalam diagnosis penyakit jantung adalah tingkat penyempitan pembuluh darah.

Namun, proses diagnosis seringkali membutuhkan pemeriksaan lanjutan yang tidak selalu cepat atau murah. Oleh karena itu, diperlukan pendekatan berbasis data untuk membantu memprediksi risiko penyakit jantung secara lebih dini menggunakan parameter yang tersedia.

#### **Tujuan**
1. Mengembangkan model klasifikasi untuk memprediksi variabel target 'num'.
2. Melakukan perbandingan beberapa algoritma machine learning, yaitu Regresi Logistik, Decision Tree, Random Forest, SVM, dan KNN.
3. Menentukan model terbaik berdasarkan performa evaluasi.

#### **Definisi Target**
Variabel target yang digunakan adalah 'num' dengan keterangan sebagai berikut:
1. '0' --> penyempitan pembuluh darah < 50%.
2. '1' --> penyempitan pembuluh darah > 50%.
Sehingga, permasalahan ini termasuk dalam binary classification.

#### **Business Objective**
1. Membantu tenaga medis dalam melakukan screening awal.
2. Mengidentifikasi pasien berisiko tinggi secara lebih cepat.
3. Mendukung pengambilan keputusan berbasis data.

#### **Manfaat**
1. Bagi Medis
   - Deteksi dini penyakit jantung
   - Mengurangi keterlambatan diagnosis
   - Membantu prioritas penanganan pasien
2. Bagi Pasien
   - Mendapatkan peringatan risiko lebih awal
   - Meningkatkan peluang pencegahan

#### **Metode Evaluasi**
1. Accuracy
2. Precision
3. Recall
4. F1-Score
5. Confusion Matrix
Dalam permasalahan ini, Recall untuk kelas positif (num = 1) menjadi fokus utama. Hal ini karena kesalahan dalam mendeteksi pasien yang benar-benar sakit (false negative) dapat berdampak serius.

## **Data Understanding**

#### **Impor Data**

In [ ]:
import pandas as pd
data_adrive = '/content/drive/MyDrive/Bootcamp, Belajar Bahasa, and Course/Celerates/Applied Machine Learning/Tugas Modul Akhir/data (1).csv'
data = pd.read_csv(data_drive)
data.head(5)

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num
0,28,1,2,130,132,0,2,185,0,0.0,?,?,?,0
1,29,1,2,120,243,0,0,160,0,0.0,?,?,?,0
2,29,1,2,140,?,0,0,170,0,0.0,?,?,?,0
3,30,0,1,170,237,0,1,170,0,0.0,?,?,6,0
4,31,0,2,100,219,0,1,150,0,0.0,?,?,?,0


In [ ]:
data.rename(columns={'num       ' : 'num'}, inplace=True)

#### **Karakteristik Dari Setiap Variabel Pada Dataset**

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       294 non-null    int64  
 1   sex       294 non-null    int64  
 2   cp        294 non-null    int64  
 3   trestbps  294 non-null    object 
 4   chol      294 non-null    object 
 5   fbs       294 non-null    object 
 6   restecg   294 non-null    object 
 7   thalach   294 non-null    object 
 8   exang     294 non-null    object 
 9   oldpeak   294 non-null    float64
 10  slope     294 non-null    object 
 11  ca        294 non-null    object 
 12  thal      294 non-null    object 
 13  num       294 non-null    int64  
dtypes: float64(1), int64(4), object(9)
memory usage: 32.3+ KB


Dari informasi di atas, dapat diketahui beberapa informasi sebagai berikut:
1. Dimensi Dataset
    - Dimensi dataset adalah (294, 14) yang berarti dataset terdiri dari 294 baris dan 14 kolom.
2. Tipe Data
    - Terdapat 4 variabel yang memiliki tipe data berupa integer, yaitu age, sex, cp, dan num.
    - Terdapat 1 variabel yang memiliki tipe data berupa float, yaitu old peak.
    - Terdapat 9 variabel yang memiliki tipe data berupa object, yaitu trestbps, chol, fbs, restecg, thalach, exang, slope, ca, dan thal.
3. Definisi Setiap Variabel
    - age: usia pasien dalam tahun
    - sex: jenis kelamin pasien (1 = male; 0 = female)
    - cp: chest pain type (1 = typical angina; 2 = atypical angina; 3 = nonanginal pain; 4 = asymptomatic)
    - trestbps: resting blood pressure (in mm Hg)
    - chol: serum cholesterol (in mg/dl)
    - fbs: fasting blood sugar > 120mg/dl (1 = true; 0 = false)
    - restecg: resting electrocardiographic results (0 = normal; 1 = having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of 0.05 mV); 2 = showing probable or definite left ventricular hypertrophy by Ester' criteria)
    - thalach: maximum heart rate achieved
    - exang: exercise induces angina/angina yg dipicu/disebabkan oleh olahraga (1 = yes; 0 = no)
    - oldpeak: ST depression induced by exercise relative to rest/penurunan segmen ST pada EKG saat olahraga dibandingkan saat istirahat
    - slope: the slope of the peak exercise ST segment/kemiringan segmen ST puncak latihan (1 = upsloping; 2 = flat; 3 = downsloping)
    - ca: number of major vessels (0--3) colored by fluoroscopy/jumlah pembuluh darah utama yang diwarnai oleh fluoroskopi
    - thal: 3 = normal; 6 = fixed defect/cacat tetap; 7 = reversable defect/cacat yang dapat dibalik
    - num: diagnosis of heart disease (angiographic disease status --> 0 = <50% diameter narrowing; 1 = >50% diameter narrowing)

#### **Pemeriksaan Imbalanced Class**

In [ ]:
counts = data['num'].value_counts()
print(counts)

num
0    188
1    106
Name: count, dtype: int64


In [ ]:
persentase = round(counts / len(data['num']), 2)
print(persentase)

num
0    0.64
1    0.36
Name: count, dtype: float64


Berdasarkan distribusi target di atas, didapatkan proporsi kelas 0 sebesar 64% dan kelas 1 sebesar 36%. Hal ini menunjukkan adanya ketidakseimbangan kelas dalam tingkat yang tergolong ringan. Namun, perbedaan proporsi tersebut tidak terlalu signifikan sehingga tidak diperlukan teknik balancing data.



## **Data Preparation**

#### **Pengubahan Tipe Data**

In [ ]:
print(data.dtypes)

age           int64
sex           int64
cp            int64
trestbps     object
chol         object
fbs          object
restecg      object
thalach      object
exang        object
oldpeak     float64
slope        object
ca           object
thal         object
num           int64
dtype: object


Tahap pertama dari Data Preparation adalah Pengubahan Tipe Data. Hal ini penting dilakukan karena adanya nilai nonnumerik, yaitu simbol "?" yang menyebabkan beberapa variabel terbaca sebagai objek.

Di sini, tahap pengubahan atau convert dibagi menjadi dua, yaitu convert objek ke numerik dan convert kategorikal ke integer.
1. Convert Objek ke Numerik
   - Tipe convert ini mengubah tipe data dari kolom yang bersifat kuantitatif menjadi numerik (float).
   - Kolom yang akan dilakukan convert menggunakan tipe ini adalah trestbps dan thalach.
2. Convert Kategorikal ke Integer
   - Tipe convert ini mengubah tipe data dari kolom kategorikal menjadi tipe integer.
   - Kolom yang akan dilakukan convert menggunakan tipe ini adalah fbs, restecg, exang, slope, ca, dan thal.

In [ ]:
#convert objek ke numerik
cols_numeric = ['trestbps', 'thalach']

for col in cols_numeric:
  data[col] = pd.to_numeric(data[col], errors='coerce')

#convert kategorikal ke int
cols_categorikal = ['fbs', 'restecg', 'exang', 'slope', 'ca', 'thal']
for col in cols_categorikal:
  data[col] = pd.to_numeric(data[col], errors='coerce').astype('Int64')

In [ ]:
print(data.dtypes)

age           int64
sex           int64
cp            int64
trestbps    float64
chol         object
fbs           Int64
restecg       Int64
thalach     float64
exang         Int64
oldpeak     float64
slope         Int64
ca            Int64
thal          Int64
num           int64
dtype: object


#### **Pemeriksaan Missing Value**

In [ ]:
data.head(20)

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num
0,28,1,2,130.0,132,0,2,185.0,0,0.0,<NA>,<NA>,<NA>,0
1,29,1,2,120.0,243,0,0,160.0,0,0.0,<NA>,<NA>,<NA>,0
2,29,1,2,140.0,?,0,0,170.0,0,0.0,<NA>,<NA>,<NA>,0
3,30,0,1,170.0,237,0,1,170.0,0,0.0,<NA>,<NA>,6,0
4,31,0,2,100.0,219,0,1,150.0,0,0.0,<NA>,<NA>,<NA>,0
5,32,0,2,105.0,198,0,0,165.0,0,0.0,<NA>,<NA>,<NA>,0
6,32,1,2,110.0,225,0,0,184.0,0,0.0,<NA>,<NA>,<NA>,0
7,32,1,2,125.0,254,0,0,155.0,0,0.0,<NA>,<NA>,<NA>,0
8,33,1,3,120.0,298,0,0,185.0,0,0.0,<NA>,<NA>,<NA>,0
9,34,0,2,130.0,161,0,0,190.0,0,0.0,<NA>,<NA>,<NA>,0


In [ ]:
data.isnull().sum()

,0
age,0
sex,0
cp,0
trestbps,1
chol,0
fbs,8
restecg,1
thalach,1
exang,1
oldpeak,0


In [ ]:
persen_question = round(data.isnull().sum() / len(data) * 100, 2)
persen_missing = persen_question[persen_question > 0]
print(persen_missing)

trestbps     0.34
fbs          2.72
restecg      0.34
thalach      0.34
exang        0.34
slope       64.63
ca          98.98
thal        90.48
dtype: float64


Dari hasil di atas, terdapat 3 kategori missing values:
1. Missing Value Rendah < 5%
   - terdapat 4 kolom yang termasuk dalam kategori ini, yaitu trestbps 0.34%, fbs 2.72%, restecg 0.34%, thalach 0.34%, dan exang 0.34%.
   - penangangan pada kategori ini adalah drop row karena kehilangan data sangat kecil dan tidak akan merusak distribusi.  
2. Missing Value Sedang 5% - 50%
   - terdapat 1 kolom yang termasuk dalam kategori ini, yaitu chol 7.82%.
   - penanganan pada kategori ini adalah imputasi karena jika dilakukan drop row maka data akan berkurang cukup besar.
3. Missing Value Tinggi > 50%
   - terdapat 3 kolom yang termasuk dalam kategori ini, yaitu slope 64.63%, thal 90.48%, dan ca 98.98%.
   - penanganan pada kategori ini adalah drop column karena variabel dengan missing > 50% tidak memberikan kontribusi signifikan pada model dan lebih baik dihapus.

Urutan penanganan missing value dimulai dari Missing Value Tinggi, lalu Missing Value Sedang, dan terakhir Missing Value Rendah.

##### **Penangangan Missing Value Tinggi (slope, thal, ca)**

In [ ]:
data = data.drop(columns=['slope', 'thal', 'ca'])

Penangangan Missing Value Tinggi dilakukan dengan cara melakukan drop column slope, thal, dan ca menggunakan fungsi drop(columns=).

##### **Penanganan Missing Value Sedang (chol)**

In [ ]:
# Penanganan Missing Value Sedang (chol)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       294 non-null    int64  
 1   sex       294 non-null    int64  
 2   cp        294 non-null    int64  
 3   trestbps  293 non-null    float64
 4   chol      294 non-null    object 
 5   fbs       286 non-null    Int64  
 6   restecg   293 non-null    Int64  
 7   thalach   293 non-null    float64
 8   exang     293 non-null    Int64  
 9   oldpeak   294 non-null    float64
 10  num       294 non-null    int64  
dtypes: Int64(3), float64(3), int64(4), object(1)
memory usage: 26.3+ KB


Untuk mencari banyaknya outlier, maka tipe data variabel chol perlu diubah terlebih dahulu dari object menjadi numerik menggunakan fungsi to_numeric.  

In [ ]:
#ubah tipe data variabel chol menjadi numerik
data['chol'] = pd.to_numeric(data['chol'], errors='coerce')

In [ ]:
chol_clean = data['chol'].dropna()

q1 = chol_clean.quantile(0.25)
q3 = chol_clean.quantile(0.75)
iqr = q3 - q1

lower_bound = q1 - 1.5 * iqr
upper_bound = q3 + 1.5 * iqr

outliers = chol_clean[(chol_clean < lower_bound) | (chol_clean > upper_bound)]

print("Nilai lower bound:", lower_bound)
print("Nilai upper bound:", upper_bound)
print("Jumlah outliers:", len(outliers))

Nilai lower bound: 98.75
Nilai upper bound: 392.75
Jumlah outliers: 11


Dari hasil di atas dapat diketahui bahwa terdapat 11 outlier. Karena terdapat outlier, maka metode imputasi yang baik untuk digunakan adalah imputasi median. Hal ini karena median robust terhadap outlier dibandingkan mean.  

In [ ]:
data['chol'] = data['chol'].fillna(data['chol'].median())

In [ ]:
print(data['chol'].isna().sum())

0


##### **Penanganan Missing Value Rendah (trestbps, fbs, restecg, thalach, dan exang)**

In [ ]:
data = data.dropna(subset=['trestbps', 'fbs', 'restecg', 'thalach', 'exang'])

In [ ]:
data.isnull().sum()

,0
age,0
sex,0
cp,0
trestbps,0
chol,0
fbs,0
restecg,0
thalach,0
exang,0
oldpeak,0


Dari hasil di atas dapat diketahui bahwa seluruh missing value telah berhasil diatasi.

#### **Pemeriksaan Duplicate Value**

In [ ]:
data.duplicated().sum()

np.int64(1)

In [ ]:
duplicate_row = data[data.duplicated(keep=False)]
print(duplicate_row)

     age  sex  cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  num
101   49    0   2     110.0  243.0    0        0    160.0      0      0.0    0
102   49    0   2     110.0  243.0    0        0    160.0      0      0.0    0


Dari hasil di atas terlihat bahwa terdapat duplicated values pada baris 101 dan 102 sehingga perlu dilakukan penanganan berupa drop salah satunya.

In [ ]:
data.drop_duplicates(inplace=True)

In [ ]:
data.duplicated().sum()

np.int64(0)

Dari hasil di atas dapat terlihat bahwa sudah tidak terdapat duplicated value.

#### **One Hot Encoding**

Selanjutnya, langkah terakhir pada tahap Data Preparation adalah melakukan encoding terhadap variabel cp dan restecg. Teknik encoding yang digunakan adalah One Hot Encoding karena kedua variabel tersebut bersifat nominal kategorikal, yaitu tidak memiliki hubungan ordinal antar kategorinya. Jadi, setiap kategori direpresentasikan sebagai variabel biner yang terpisah agar dapat diolah dengan baik oleh model machine learning.

In [ ]:
data_encoding = pd.get_dummies(data, columns=['cp','restecg'], drop_first=True)

In [ ]:
data_encoding.head()

,age,sex,trestbps,chol,fbs,thalach,exang,oldpeak,num,cp_2,cp_3,cp_4,restecg_1,restecg_2
0,28,1,130.0,132.0,0,185.0,0,0.0,0,True,False,False,False,True
1,29,1,120.0,243.0,0,160.0,0,0.0,0,True,False,False,False,False
2,29,1,140.0,243.0,0,170.0,0,0.0,0,True,False,False,False,False
3,30,0,170.0,237.0,0,170.0,0,0.0,0,False,False,False,True,False
4,31,0,100.0,219.0,0,150.0,0,0.0,0,True,False,False,True,False


In [ ]:
data_encoding.columns

Index(['age', 'sex', 'trestbps', 'chol', 'fbs', 'thalach', 'exang', 'oldpeak',
       'num', 'cp_2', 'cp_3', 'cp_4', 'restecg_1', 'restecg_2'],
      dtype='object')

## **Modeling**

#### **Split Data**

In [ ]:
from sklearn.model_selection import train_test_split, GridSearchCV

x = data_encoding.drop('num', axis = 1)
y = data_encoding['num']

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 42, stratify = y)

Langkah pertama dalam tahap modeling adalah melakukan split data menjadi dua bagian, yaitu data training dan data testing. Data training mencakup seluruh variabel kecuali variabel 'num'. Sementara itu, data testing hanya mencakup variabel 'num' saja.

Proporsi antara data training dan data testing menggunakan perbandingan 80% : 20% dengan penggunaan stratify = y agar seimbang antara kelas 0 dan kelas 1.

#### **Scaling**

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

x_train_scaled = scaler.fit_transform(x_train)
x_test_scaler = scaler.transform(x_test)

Scaling diperlukan dalam model Regresi Logistik, SVM dan KNN. Regresi Logistik membutuhkan scaling karena menggunakan jarak (distance). SVM perlu scaling karena menggunakan margin dan distance. Regresi Logistik menggunakan scaling karena menggunakan optimasi (mencari bobot terbaik dengan meminimalkan error pakai iterasi (gradient descent)).

In [ ]:
from sklearn.pipeline import Pipeline

#### **Regresi Logistik**

In [ ]:
from sklearn.linear_model import LogisticRegression

pipe_rl = Pipeline([
    ('scaler', scaler),
    ('model', LogisticRegression(max_iter=1000))
])

param_rl = {
    'model__C': [0.01, 0.1, 1, 10],
    'model__penalty': ['l2'],
    'model__solver': ['lbfgs']
}

grid_rl = GridSearchCV(pipe_rl, param_rl, cv=5, scoring='recall')
grid_rl.fit(x_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('model',
                                        LogisticRegression(max_iter=1000))]),
             param_grid={'model__C': [0.01, 0.1, 1, 10],
                         'model__penalty': ['l2'], 'model__solver': ['lbfgs']},
             scoring='recall')

#### **Decision Tree**

In [ ]:
from sklearn.tree import DecisionTreeClassifier

pipe_dt = Pipeline([
    ('model', DecisionTreeClassifier(random_state=42))
])

param_dt = {
    'model__max_depth': [3, 5, 10, None],
    'model__min_samples_split': [2, 5, 10],
    'model__min_samples_leaf': [1,2,4]
}

grid_dt = GridSearchCV(pipe_dt, param_dt, cv=5, scoring='recall')
grid_dt.fit(x_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('model',
                                        DecisionTreeClassifier(random_state=42))]),
             param_grid={'model__max_depth': [3, 5, 10, None],
                         'model__min_samples_leaf': [1, 2, 4],
                         'model__min_samples_split': [2, 5, 10]},
             scoring='recall')

#### **Random Forest**

In [ ]:
from sklearn.ensemble import RandomForestClassifier

pipe_rf = Pipeline([
    ('model', RandomForestClassifier(random_state=42))
])

param_rf = {
    'model__n_estimators': [100, 200],
    'model__max_depth': [5, 10, None],
    'model__min_samples_split': [2, 5],
    'model__min_samples_leaf': [1, 2]
}

grid_rf = GridSearchCV(pipe_rf, param_rf, cv=5, scoring='recall')
grid_rf.fit(x_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('model',
                                        RandomForestClassifier(random_state=42))]),
             param_grid={'model__max_depth': [5, 10, None],
                         'model__min_samples_leaf': [1, 2],
                         'model__min_samples_split': [2, 5],
                         'model__n_estimators': [100, 200]},
             scoring='recall')

#### **Support Vector Machine (SVM)**

In [ ]:
from sklearn.svm import SVC

pipe_svm = Pipeline ([
    ('scaler', scaler),
    ('model', SVC())
])

param_svm = {
    'model__C': [0.1, 1, 10],
    'model__kernel': ['linear', 'rbf'],
    'model__gamma': ['scale', 'auto']
}

grid_svm = GridSearchCV(pipe_svm, param_svm, cv=5, scoring='recall')
grid_svm.fit(x_train, y_train)


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('model', SVC())]),
             param_grid={'model__C': [0.1, 1, 10],
                         'model__gamma': ['scale', 'auto'],
                         'model__kernel': ['linear', 'rbf']},
             scoring='recall')

#### **K-Nearest Neighbor (KNN)**

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

pipe_knn = Pipeline([
    ('scaler', scaler),
    ('model', KNeighborsClassifier())
])

param_knn = {
    'model__n_neighbors': [3, 5, 7, 9],
    'model__weights': ['uniform', 'distance'],
    'model__metric': ['euclidean', 'manhattan']
}

grid_knn = GridSearchCV(pipe_knn, param_knn, cv=5, scoring='recall')
grid_knn.fit(x_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('scaler', StandardScaler()),
                                       ('model', KNeighborsClassifier())]),
             param_grid={'model__metric': ['euclidean', 'manhattan'],
                         'model__n_neighbors': [3, 5, 7, 9],
                         'model__weights': ['uniform', 'distance']},
             scoring='recall')

## **Evaluation**

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

results = {}

models = {
    'Regresi Logistik': grid_rl,
    'Decision Tree': grid_dt,
    'Random Forest': grid_rf,
    'SVM': grid_svm,
    'KNN': grid_knn
}

for name, model in models.items():
  best_model = model.best_estimator_
  y_pred = best_model.predict(x_test)

  results[name] = {
      'Best Parameter': model.best_params_,
      'Accuracy': accuracy_score(y_test, y_pred),
      'Precision': precision_score(y_test, y_pred),
      'Recall': recall_score(y_test, y_pred),
      'F1 Score': f1_score(y_test, y_pred)
  }

  hasil = pd.DataFrame(results).T

In [ ]:
hasil

,Best Parameter,Accuracy,Precision,Recall,F1 Score
Regresi Logistik,"{'model__C': 0.1, 'model__penalty': 'l2', 'mod...",0.859649,0.842105,0.761905,0.8
Decision Tree,"{'model__max_depth': 3, 'model__min_samples_le...",0.736842,0.666667,0.571429,0.615385
Random Forest,"{'model__max_depth': None, 'model__min_samples...",0.77193,0.7,0.666667,0.682927
SVM,"{'model__C': 0.1, 'model__gamma': 'scale', 'mo...",0.842105,0.833333,0.714286,0.769231
KNN,"{'model__metric': 'manhattan', 'model__n_neigh...",0.824561,0.761905,0.761905,0.761905


--> Dari hasil di atas, dapat diketahui bahwa model terbaik adalah Regresi Logistik karena memiliki skor recall tertinggi, yaitu 76.19%. Hal ini menunjukkan bahwa model berhasil mengenali sekitar 76.19% pasien yang benar-benar memiliki penyempitan pembuluh darah signifikan (>50%), yaitu pasien dengan label 1, yakni terindikasi penyakit jantung.

Dengan kata lain, model masih belum berhasil mendeteksi seluruh pasien berisiko karena sekitar 23.81% pasien dengan kondisi penyempitan pembuluh darah >50% tidak dikenali oleh model (false negative).

--> Model Regresi Logistik terbaik adalah menggunakan kekuatan regularisasi atau C sebesar 0.1, dengan regularisasi yang digunakan ialah L2 atau Ridge, lalu solvernya adalah lbfgs.   

#### **Feature Importance (Koefisien)**

In [ ]:
import numpy as np

best_rl = grid_rl.best_estimator_

coef = best_rl.named_steps['model'].coef_[0]

feature_importance_rl = pd.DataFrame({
    'Fitur': x.columns,
    'Koefisien': coef
}).sort_values(
    by='Koefisien',
    ascending=False
)

print(feature_importance_rl)



        Fitur  Koefisien
7     oldpeak   0.604027
10       cp_4   0.503405
6       exang   0.463899
1         sex   0.462749
4         fbs   0.343552
3        chol   0.265577
0         age   0.037633
2    trestbps  -0.029298
9        cp_3  -0.074465
12  restecg_2  -0.080052
11  restecg_1  -0.136105
5     thalach  -0.172324
8        cp_2  -0.459513


Dari data di atas, dapat diambil beberapa informasi awal sebagai berikut:
1. Koefisien + berarti akan menaikkan peluang kelas 1
2. Koefisien - berarti akan menurunkan peluang kelas 1

Kemudian, dapat diambil informasi lanjutan untuk setiap fiturnya yakni sebagai berikut:
1. Fitur yang paling berpengaruh dalam meningkatkan risiko kelas 1 adalah oldpeak dengan koefisien 0.604. Semakin tinggi nilai oldpeak atau indikasi jantung bekerja tidak normal saat olahraga, maka probabilitas pasien masuk ke kelas 1 semakin meningkat.
2. Fitur cp_2 atau atypical angina memiliki koefisien negatif sebesar -0.459 yang mana menunjukkan bahwa keberadaan chest pain tipe atypical angina menurunkan probablitas pasien terklasifikasi sebagai kelas 1 atau penyempitan pembuluh darah > 50% dibandingkan kategori referensi (cp_1) atau typical angina.

#### **Classification Report**

In [ ]:
from sklearn.metrics import classification_report

y_pred_lr = best_rl.predict(x_test)

report_lr = classification_report(y_test, y_pred_lr)
print("Classification Report - Regresi Logistik:\n")
print(report_lr)

Classification Report - Regresi Logistik:

              precision    recall  f1-score   support

           0       0.87      0.92      0.89        36
           1       0.84      0.76      0.80        21

    accuracy                           0.86        57
   macro avg       0.86      0.84      0.85        57
weighted avg       0.86      0.86      0.86        57



--> Nilai accuracy, precision, recall, f1-score, dan support di atas berbeda hasilnya dengan yang dihasilkan oleh yang sebelumnya karena sebelumnya menggunakan GridSearchCV dimana data berasal dari banyak fold (CV), sedangkan pada classification report data hanya berasal dari 1 test split saja.

--> Pada classification report di atas terlihat bahwa kelas 1 atau pasien dengan penyempitan pembuluh darah > 50% sebanyak 21 orang, sedangkan kelas 0 atau pasien dengan penyempitan pembuluh darah < 50% sebanyak 36 orang.

--> Nilai recall yang cukup tinggi, yakni sebesar 92% menunjukkan bahwa model berhasil mengenali pasien dengan penyempitan pembuluh darah > 50% dengan nyaris sempurna dimana hanya 8% pasien saja yang tidak dikenali oleh model.  

## **Uji Coba Data Baru**

In [ ]:
import pandas as pd

user_data = pd.DataFrame([{
    "age": 55,
    "sex": 1,
    "cp": 3,
    "trestbps": 140,
    "chol": 250,
    "fbs": 0,
    "restecg": 1,
    "thalach": 150,
    "exang": 0,
    "oldpeak": 1.2,
    "slope": 2,
    "ca": 1,
    "thal": 3
}])

user_encoded = pd.get_dummies(user_data).reindex(columns=x.columns, fill_value=0) #semua kolom training wajib ada, yang tidak ada akan diisi 0

prediction = best_model.predict(user_encoded)
print("Hasil prediksi:", prediction)

Hasil prediksi: [0]


In [ ]:
import pandas as pd

user_data = pd.DataFrame([{
    "age": 55,
    "sex": 1,
    "cp": 3,
    "trestbps": 140,
    "chol": 250,
    "fbs": 0,
    "restecg": 1,
    "thalach": 150,
    "exang": 0,
    "oldpeak": 2.5,
    "slope": 2,
    "ca": 1,
    "thal": 3
}])

user_encoded = pd.get_dummies(user_data).reindex(columns=x.columns, fill_value=0)

prediction = best_model.predict(user_encoded)
print("Hasil prediksi:", prediction)

Hasil prediksi: [1]


Dari dua hasil uji coba di atas dapat dilihat bahwa semakin tinggi nilai oldpeak, maka akan probabilitas pasien masuk ke kelas 1 akan semakin meningkat.